Support setting Iceberg table format version

alexjo2144 opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->

## Description

Allows setting the table format for newly created Iceberg tables. Defaults to v1.

<!-- Elaborate beyond the title of the PR as necessary to help the reviewers and maintainers.-->

<!-- Answer the following questions to help reviewers and maintainers
understand this PR's scope at a glance:
-->

> Is this change a fix, improvement, new feature, refactoring, or other?

Feature

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

Iceberg connector

> How would you describe this change to a non-technical end user or system administrator?

Allows creating either v1 or v2 Iceberg tables

## Related issues, pull requests, and links

<!-- List any issues fixed by this PR, and provide links to other related PRs, upstream release notes, and other useful resources. For example:
* Fixes #issuenumber
* Related documentation in #issuenumber
* [Some release notes](http://usefulinfo.example.com)
-->

<!-- The following sections are filled in by the maintainer with input from the contributor:
Use :white_check_mark: or (x) to signal selection.
-->

## Documentation

( ) No documentation is needed.
(x) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# Iceberg
* Support creating tables with either Iceberg format version 1, or 2.
```

findinpath commented on docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -441,6 +441,10 @@ Property Name                                      Description
 ``location``                                       Optionally specifies the file system location URI for
                                                    the table.
+
+``format_version``                                 Optionally specifies the format version of the Iceberg
```
From a new user perspective I think it would be useful to have a very high-level description about what `1` and `2` actually mean.

findinpath commented on docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -441,6 +441,10 @@ Property Name                                      Description
                                                    the table.
+
+``format_version``                                 Optionally specifies the format version of the Iceberg
+                                                   specification to use for new tables; either ``1`` or ``2``.
```
to use for tables created by Trino

findinpath commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergTableProperties.java:
```diff
@@ -64,6 +71,11 @@ public IcebergTableProperties(IcebergConfig icebergConfig)
                         false))
+                .add(integerProperty(
+                        FORMAT_VERSION_PROPERTY,
+                        "Iceberg table format version for newly created tables",
```
for tables created by Trino

findepi replied:
this applies to CREATE TABLE, but also to SHOW CREATE, so should just describe what it is:

"Iceberg table format version"

findepi commented:
What's the benefit?

Until v2 is fully supported, i don't think we should merge this.

alexjo2144 commented:
> What's the benefit?

There are some spec changes in v2 besides merge-on-read row level deletes. https://iceberg.apache.org/spec/#version-2 A user might want v2 tables for those reasons, but we can combine this with https://github.com/trinodb/trino/pull/11886 if you'd like

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergConfig.java:
```diff
@@ -40,6 +44,7 @@
     private boolean tableStatisticsEnabled = true;
     private boolean projectionPushdownEnabled = true;
     private Optional<String> hiveCatalogName = Optional.empty();
+    private int formatVersion = FORMAT_VERSION_SUPPORT_MIN;
```
This is compatible, but please follow up to create v2 by default.
We are implementing v2 deletes, and are currently not working on copy-on-write delete mode (as required for v1), so v2 are better for end users.

kbendick replied:
If `copy-on-write` isn't implemented, do make sure that `write.format.default` is `merge-on-read` then. Particularly for users who might interact with the table via other engines.

kbendick replied:
Actually there are 3 configurations coming up with respect to `copy-on-write` vs `merge-on-read`:

https://github.com/apache/iceberg/blob/a2392355e2a9afe757b700409882f26737a5062e/docs/tables/configuration.md?plain=1#L75-L79

alexjo2144 replied:
Yep, at the moment neither merge-on-read or copy-on-write are supported so I didn't add those three options in here, but they are in the PR adding support for merge-on-read deletes: https://github.com/trinodb/trino/pull/11886/commits/8ebba99d0642db3b6b4fdddfa3aa449c369cd8be#diff-2af3e19a6b656640a7d0bb73114ef224953a2efa04e569b1fe4da953b2cc6d15R411

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergTableProperties.java:
```diff
@@ -64,6 +71,11 @@ public IcebergTableProperties(IcebergConfig icebergConfig)
+                        FORMAT_VERSION_PROPERTY,
+                        "Iceberg table format version for newly created tables",
+                        icebergConfig.getFormatVersion(),
+                        false))
```
add validation that it's between min and max supported

(`Consumer<Integer> validation`)

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergTableProperties.java:
```diff
@@ -88,4 +100,14 @@ public static Optional<String> getTableLocation(Map<String, Object> tablePropert
+    public static int getFormatVersion(Map<String, Object> tableProperties)
+    {
+        int version = (int) tableProperties.get(FORMAT_VERSION_PROPERTY);
+        if (version < FORMAT_VERSION_SUPPORT_MIN || version > FORMAT_VERSION_SUPPORT_MAX) {
```
Move the validation to property definition.

findepi reviewed (approved):
Should there be a test?

alexjo2144 commented:
@findepi addressed comments, @findinpath added to the docs a little, explaining that v2 is required for row level deletes and linking the Iceberg documentation

findepi commented on docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -441,6 +441,12 @@ Property Name                                      Description
+                                                   specification to use for new tables; either ``1`` or ``2``.
+                                                   Defaults to ``1``. Version 2 is required for row level deletes.
+                                                   The full list of changes from v1 to v2 are outlined
+                                                   `here <https://iceberg.apache.org/spec/#version-2-row-level-deletes>`_.
```
https://iceberg.apache.org/spec/#version-2 would be a correct link.
However, i am fine with omitting the whole sentence, we can consider this out of scope of this document.

From Trino user perspective, we should say that "2" is required for DML.

alexjo2144 replied:
> From Trino user perspective, we should say that "2" is required for DML

I was going to put that, but not until after DML is in

alexjo2144 replied:
Removed the last sentence

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergConfig.java:
```diff
@@ -182,4 +187,19 @@ public IcebergConfig setHiveCatalogName(String hiveCatalogName)
+    }
+
+    @Config("iceberg.format-version")
+    @ConfigDescription("Iceberg table format version")
```
"Default ..."
